afqinsight/CovariateRegressor.py

```diff
@@ -0,0 +1,225 @@
+            estimated from the train-set to the test set (cross_validate=True)
+            or whether to fit the covariate regressor separately on the test-set
+            (cross_validate=False). Setting this parameter to True is equivalent
+            to "foldwise covariate regression" (FwCR) as described in our paper
+            (https://www.biorxiv.org/content/early/2018/03/28/290684). Setting
+            this parameter to False, however, is NOT equivalent to "whole
+            dataset covariate regression" (WDCR) as it does not apply covariate
+            regression to the *full* dataset, but simply refits the covariate
+            model on the test-set. We recommend setting this parameter to True.
+        precise: bool
+            Transformer-objects in scikit-learn only allow to pass the data
+            (X) and optionally the target (y) to the fit and transform methods.
+            However, we need to index the covariate accordingly as well. To do so,
+            we compare the X during initialization (self.X) with the X passed to
+            fit/transform. As such, we can infer which samples are passed to the
+            methods and index the covariate accordingly. When setting precise to
+            True, the arrays are compared feature-wise, which is accurate, but
+            relatively slow. When setting precise to False, it will infer the index
+            by looking at the hash of all the features, which is much
+            faster. Also, to aid the accuracy, we remove the features which are constant
+            (0) across samples.
+        stack_intercept : bool
+            Whether to stack an intercept to the covariate (default is True)
+
+        Attributes
+        ----------
+        weights_ : numpy array
+            Array with weights for the covariate(s).
+        """
+        self.covariate = covariate.astype(np.float64)
+        self.cross_validate = cross_validate
+        self.X = X
+        self.precise = precise
+        self.stack_intercept = stack_intercept
+        self.weights_ = None
+        self.pipeline = pipeline
+        self.imputer = SimpleImputer(strategy="median")
+        self.X_imputer = SimpleImputer(strategy="median")
+        self.unique_id_col_index = unique_id_col_index
+
+    def _prepare_covariate(self, covariate):
+        """Prepare covariate matrix (adds intercept if needed)"""
+        if self.stack_intercept:
+            return np.c_[np.ones((covariate.shape[0], 1)), covariate]
+        return covariate
+
+    def fit(self, X, y=None):
+        """Fits the covariate-regressor to X.
+
+        Parameters
+        ----------
+        X : numpy array
+            An array of shape (n_samples, n_features), which should correspond
+            to your train-set only!
+        y : None
+            Included for compatibility; does nothing.
+        """
+
+        # Prepare covariate matrix (adds intercept if needed)
+        covariate = self._prepare_covariate(self.covariate)
```

Comment:
I think you can replace this with and get rid of the `_prepare_covariate` method, which is essentially just this if statement:
```suggestion
        if self.stack_intercept:
            covariate = np.c_[np.ones((covariate.shape[0], 1)), covariate]
```

Or am I missing something?

Reply:
We repeat this in both fit and transform, so we thought it was a good idea to keep this as a separate DRY function.